llvm/lib/Transforms/Scalar/LICM.cpp

```diff
@@ -2719,6 +2728,65 @@ static bool hoistFPAssociation(Instruction &I, Loop &L,
   return true;
 }
 
+static bool hoistIntAssociation(Instruction &I, Loop &L,
```

Comment:
Comment as much on the existing code as the new code.  We're going through the following steps:
0:  ((A1 * B1) + (A2 * B2) + ...) * C where
1: ((A1 * C * B1) + (A2 * C * B2) + ...) 
2: ((A1 * C) * B1 + (A2 * C) * B2 + ...) 

Isn't the step going from 0 to 1 distribution, not associativity?  If so, some comments need clarified.

Reply:
Actually, if I'm right here about there being a distribution step here, isn't the prior FP code unsound?  At least according to wikipedia (https://en.wikipedia.org/wiki/Distributive_property), floating point multiplication does not distribute over addition, and I don't know hasAllowReassoc gives enough freedom.  (Since this isn't re-association).

Note that 2s complement integer math doesn't have this problem.

Reply:
This should probably be explicitly mentioned in LangRef, but we widely assume that reassoc also permits distribution.